src/cls/rbd/cls_rbd.cc

```diff
@@ -5920,6 +5921,56 @@ int mirror_mode_set(cls_method_context_t hctx, bufferlist *in,
     if (r < 0) {
       return r;
     }
+
+    r = remove_key(hctx, mirror::REMOTE_NAMESPACE);
```

Comment:
I wonder if removing this key here is still desirable.  Since we went with separate `mirror_remote_namespace_get()`/`mirror_remote_namespace_set()` APIs and `mirror_mode()` API isn't touched, we end up with behavior that might be considered surprising:
```
< operating on ns-local ioctx, mirroring is disabled initially >
mirror_remote_namespace_get()  --> "ns-local"
mirror_mode_set(RBD_MIRROR_MODE_IMAGE)
...
mirror_mode(RBD_MIRROR_MODE_DISABLED)
mirror_remote_namespace_get()  --> "ns-local" -- OK, didn't change
...
mirror_remote_namespace_set("ns-remote")
...
mirror_remote_namespace_get()  --> "ns-remote" -- OK, changed to reflect the set
mirror_mode_set(RBD_MIRROR_MODE_IMAGE)
...
mirror_mode(RBD_MIRROR_MODE_DISABLED)
mirror_remote_namespace_get()  --> "ns-local" -- changed, no longer reflects the set
```
Remote namespace can be set long before mirroring is enabled (i.e. `mirror_remote_namespace_set()` and `mirror_mode_set()` calls don't have to be issued one after another), so I'm having a hard time justifying to myself not preserving remote namespace across mirroring getting disabled.

Reply:
I think it is better to remove the key when the namespace disabled. Disabling a mirror namespace can be assumed to clear all previous behaviour. This is the exact reason I wanted everything done in a single API instead of a 2 step process - it is not optimal design and causes surprising behaviour.

Reply:
Since we went with a separate API, the question to answer is which would be more expected from an API user's (not the CLI user's -- there it's wrapped into a single command with the default value for the namespace, so the behavior would be the same no matter what is decided here) perspective.  Personally, if I encountered such a two-step API, I would be surprised by the namespace getting changed underneath me without a call to the setter and would rather expect it to be preserved, especially given that the site name gets preserved (when mirroring gets disabled on a pool).

But I don't have a strong opinion and can see the argument being made both ways.  If you are convinced that it's better to reset the namespace, please add a comment on `rbd_mirror_remote_namespace_set()` declaration in `src/include/rbd/librbd.h` clarifying this.